// @gnosis.pm/safe-core-sdk/dist/src/ethereumLibs/Web3Adapter.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Web3Adapter_web3, _Web3Adapter_signerAddress;
Object.defineProperty(exports, "__esModule", { value: true });
const bignumber_1 = require("@ethersproject/bignumber");
const contractInstancesWeb3_1 = require("../contracts/contractInstancesWeb3");
const safeDeploymentContracts_1 = require("../contracts/safeDeploymentContracts");
class Web3Adapter {
    constructor({ web3, signerAddress }) {
        _Web3Adapter_web3.set(this, void 0);
        _Web3Adapter_signerAddress.set(this, void 0);
        if (!web3) {
            throw new Error('web3 property missing from options');
        }
        __classPrivateFieldSet(this, _Web3Adapter_web3, web3, "f");
        __classPrivateFieldSet(this, _Web3Adapter_signerAddress, signerAddress, "f");
    }
    isAddress(address) {
        return __classPrivateFieldGet(this, _Web3Adapter_web3, "f").utils.isAddress(address);
    }
    async getBalance(address) {
        return bignumber_1.BigNumber.from(await __classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.getBalance(address));
    }
    async getChainId() {
        return __classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.getChainId();
    }
    getSafeContract({ safeVersion, chainId, isL1SafeMasterCopy, customContractAddress }) {
        const safeSingletonDeployment = (0, safeDeploymentContracts_1.getSafeContractDeployment)(safeVersion, chainId, isL1SafeMasterCopy);
        const contractAddress = customContractAddress !== null && customContractAddress !== void 0 ? customContractAddress : safeSingletonDeployment === null || safeSingletonDeployment === void 0 ? void 0 : safeSingletonDeployment.networkAddresses[chainId];
        if (!contractAddress) {
            throw new Error('Invalid Safe Proxy contract address');
        }
        const safeContract = this.getContract(contractAddress, safeSingletonDeployment === null || safeSingletonDeployment === void 0 ? void 0 : safeSingletonDeployment.abi);
        return (0, contractInstancesWeb3_1.getSafeContractInstance)(safeVersion, safeContract);
    }
    getMultiSendContract(safeVersion, chainId, customContractAddress) {
        const multiSendDeployment = (0, safeDeploymentContracts_1.getMultiSendContractDeployment)(safeVersion, chainId);
        const contractAddress = customContractAddress !== null && customContractAddress !== void 0 ? customContractAddress : multiSendDeployment === null || multiSendDeployment === void 0 ? void 0 : multiSendDeployment.networkAddresses[chainId];
        if (!contractAddress) {
            throw new Error('Invalid Multi Send contract addresss');
        }
        const multiSendContract = this.getContract(contractAddress, multiSendDeployment === null || multiSendDeployment === void 0 ? void 0 : multiSendDeployment.abi);
        return (0, contractInstancesWeb3_1.getMultiSendContractInstance)(safeVersion, multiSendContract);
    }
    getSafeProxyFactoryContract(safeVersion, chainId, customContractAddress) {
        const proxyFactoryDeployment = (0, safeDeploymentContracts_1.getSafeProxyFactoryContractDeployment)(safeVersion, chainId);
        const contractAddress = customContractAddress !== null && customContractAddress !== void 0 ? customContractAddress : proxyFactoryDeployment === null || proxyFactoryDeployment === void 0 ? void 0 : proxyFactoryDeployment.networkAddresses[chainId];
        if (!contractAddress) {
            throw new Error('Invalid Safe Proxy Factory contract address');
        }
        const proxyFactoryContract = this.getContract(contractAddress, proxyFactoryDeployment === null || proxyFactoryDeployment === void 0 ? void 0 : proxyFactoryDeployment.abi);
        return (0, contractInstancesWeb3_1.getGnosisSafeProxyFactoryContractInstance)(safeVersion, proxyFactoryContract);
    }
    getContract(address, abi) {
        return new (__classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.Contract)(abi, address);
    }
    async getContractCode(address) {
        return __classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.getCode(address);
    }
    async getTransaction(transactionHash) {
        return __classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.getTransaction(transactionHash);
    }
    async getSignerAddress() {
        return __classPrivateFieldGet(this, _Web3Adapter_signerAddress, "f");
    }
    signMessage(message) {
        return __classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.sign(message, __classPrivateFieldGet(this, _Web3Adapter_signerAddress, "f"));
    }
    estimateGas(transaction, options) {
        return __classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.estimateGas(transaction, options);
    }
    call(transaction) {
        return __classPrivateFieldGet(this, _Web3Adapter_web3, "f").eth.call(transaction);
    }
}
_Web3Adapter_web3 = new WeakMap(), _Web3Adapter_signerAddress = new WeakMap();
exports.default = Web3Adapter;
